lib/src/util/unrelated_types_visitor.dart

```diff
@@ -47,46 +67,156 @@ abstract class UnrelatedTypesProcessors extends SimpleAstVisitor<void> {
     // We've completed the "cheap" checks, and must now continue with the
     // arduous task of determining whether the method target implements
     // [definition].
+    var targetType = _getTargetType(node, node.realTarget);
+    if (targetType is! InterfaceType) {
+      return;
+    }
 
-    DartType? targetType;
-    var target = node.realTarget;
-    if (target != null) {
-      targetType = target.staticType;
-    } else {
-      for (AstNode? parent = node; parent != null; parent = parent.parent) {
-        if (parent is ClassDeclaration) {
-          targetType = parent.declaredElement?.thisType;
-          break;
-        } else if (parent is MixinDeclaration) {
-          targetType = parent.declaredElement?.thisType;
-          break;
-        } else if (parent is EnumDeclaration) {
-          targetType = parent.declaredElement?.thisType;
-          break;
-        } else if (parent is ExtensionDeclaration) {
-          targetType = parent.extendedType.type;
-          break;
-        }
+    for (var methodDefinition in matchingMethods) {
+      var collectionType = methodDefinition.collectionTypeFor(targetType);
+      if (collectionType != null) {
+        _checkMethod(node.argumentList.arguments.first, methodDefinition,
+            collectionType);
+        return;
       }
     }
+  }
 
-    if (targetType is! InterfaceType) {
-      return;
+  DartType? _getTargetType(Expression node, Expression? target) {
+    if (target != null) {
+      return target.staticType;
     }
 
-    var collectionType = targetType.asInstanceOf(interface);
-    if (collectionType == null) {
-      return;
+    for (AstNode? parent = node; parent != null; parent = parent.parent) {
+      if (parent is ClassDeclaration) {
+        return parent.declaredElement?.thisType;
+      } else if (parent is MixinDeclaration) {
+        return parent.declaredElement?.thisType;
+      } else if (parent is EnumDeclaration) {
+        return parent.declaredElement?.thisType;
+      } else if (parent is ExtensionDeclaration) {
+        return parent.extendedType.type;
+      }
     }
+    return null;
+  }
 
+  /// Checks a [MethodInvocation] or [IndexExpression] which has a singular
+  /// [argument] and matches [methodDefinition], with a target with a static
+  /// type of [collectionType].
+  void _checkMethod(Expression argument, MethodDefinition methodDefinition,
+      InterfaceType collectionType) {
     // Finally, determine whether the type of the argument is related to the
     // type of the method target.
-    var argumentType = node.argumentList.arguments.first.staticType;
+    var argumentType = argument.staticType;
+    if (argumentType == null) return;
+
+    switch (methodDefinition.expectedArgumentKind) {
```

Comment:
It's possible... I think it would actually add complexity. There are a few particulars that we'd have to push into MethodDefinition like:

* it would also need the TypeProvider, only for the Set.containsAll, Set.removeAll, and Set.retainAll cases,
* it would still need to be this big switch based on ExpectedArgumentKind; I don't think we could remove it.